code/controllers/subsystem/sounds.dm

```diff
@@ -0,0 +1,165 @@
+	var/channel_reserve_high
+
+/datum/controller/subsystem/sounds/Initialize()
+	setup_available_channels()
+	return ..()
+
+/**
+  * Sets up all available sound channels
+  */
+/datum/controller/subsystem/sounds/proc/setup_available_channels()
+	channel_list = list()
+	reserved_channels = list()
+	using_channels = list()
+	using_channels_by_datum = list()
+	for(var/i in 1 to using_channels_max)
+		channel_list += i
+	channel_random_low = 1
+	channel_reserve_high = length(channel_list)
+
+/**
+  * Removes a channel from using list
+  *
+  * Arguments:
+  * * channel - The channel number
+  */
+/datum/controller/subsystem/sounds/proc/free_sound_channel(channel)
+	var/text_channel = num2text(channel)
+	var/using = using_channels[text_channel]
+	using_channels -= text_channel
+	if(using != TRUE)		// datum channel
+		using_channels_by_datum[using] -= channel
+		if(!length(using_channels_by_datum[using]))
+			using_channels_by_datum -= using
+	free_channel(channel)
+
+/**
+  * Frees all the channels a datum is using
+  *
+  * Arguments:
+  * * D - The datum
+  */
+/datum/controller/subsystem/sounds/proc/free_datum_channels(datum/D)
+	var/list/L = using_channels_by_datum[D]
+	if(!L)
+		return
+	for(var/channel in L)
+		using_channels -= num2text(channel)
+		free_channel(channel)
+	using_channels_by_datum -= D
+
+/**
+  * Frees all datumless channels
+  */
+/datum/controller/subsystem/sounds/proc/free_datumless_channels()
+	free_datum_channels(DATUMLESS)
+
+/**
+  * NO AUTOMATIC CLEANUP - If you use this, you better manually free it later!
+  *
+  * Returns an integer for channel
+  */
+/datum/controller/subsystem/sounds/proc/reserve_sound_channel_datumless()
+	. = reserve_channel()
+	if(!.) // oh no..
+		return FALSE
+	var/text_channel = num2text(.)
+	using_channels[text_channel] = DATUMLESS
+	LAZYADD(using_channels_by_datum[DATUMLESS], .)
+
+/**
+  * Reserves a channel for a datum. Automatic cleanup only when the datum is deleted.
+  *
+  * Returns an integer for channel
+  * Arguments:
+  * * D - The datum
+  */
+/datum/controller/subsystem/sounds/proc/reserve_sound_channel(datum/D)
+	if(!D) // i don't like typechecks but someone will fuck it up
+		CRASH("Attempted to reserve sound channel without datum using the managed proc.")
+	. = reserve_channel()
+	if(!.)
+		return FALSE
+	var/text_channel = num2text(.)
+	using_channels[text_channel] = D
+	LAZYADD(using_channels_by_datum[D], .)
+
+/**
+  * Reserves a channel and updates the datastructure. Private proc.
+  */
+/datum/controller/subsystem/sounds/proc/reserve_channel()
+	PRIVATE_PROC(TRUE)
+	if(channel_reserve_high <= random_channels_min)		// out of channels
+		return
+	var/channel = channel_list[channel_reserve_high]
+	reserved_channels[num2text(channel)] = channel_reserve_high--
+	return channel
+
+/**
+  * Frees a channel and updates the datastructure. Private proc.
+  */
+/datum/controller/subsystem/sounds/proc/free_channel(number)
+	PRIVATE_PROC(TRUE)
```

Comment:
For anyone wondering, this doesnt make the proc itself private (BYOND isnt smart enough for that), but will make the linter cry out in agony if you try use this proc outside of the datum.